### PR TITLE
code style implementation

### PR DESCRIPTION
**Pull request by decanTyme**
**PR Includes**:
* Ignore .vscode folder
* Add Prettier, config, ignore, scripts
* Add ESLint, config, ignore, scripts
* Add Yarn PnP VSCode integration
* Update CONTRIBUTING.md
* Add CI workflow
* Some codestyle fixes

**Summary**:

Adds some checks and automations for anyone looking to contribute code (no testing yet.) The ESLint config extends the [eslint-config-google](https://github.com/google/eslint-config-google). Overwritten rules are of my own personal preference. 

Also, since [Yarn PnP](https://yarnpkg.com/features/pnp) is used, some contributors that might be using VSCode with _Prettier_ and _ESLint_ extensions installed might encounter issues such as failing to find their respective modules so I [generated some editor SDKs for integration with VSCode](https://yarnpkg.com/getting-started/editor-sdks#vscode). These extensions should continue to work within this workspace. Basically, they fail to find a `node_modules` folder so they error out without this.

Lastly, I don't know much about this since I've only just recently migrated to Yarn [v1], but the `.pnp.cjs` file's `packageLocation` props have changed to my own dirs (e.g., from `./.yarn/cache/...` to something like `../../../../../C:/Users/<USER>/AppData/Local/Yarn/Berry/cache/`, is that what it's supposed to do?) since I did a `yarn install`. Won't start without running it so I had no choice.

**p.s.** Reminded me of those awesome ASCII rendering stuff, but this time with colors! So got a little excited and decided to help a little. Please do tell if there were some issues. Might add more PRs for new features/tests soon.

### CODE REVIEW

**Comment by decanTyme** [comment]
I think I get why it got changed

**Comment by decanTyme** [comment]
> Lastly, I don't know much about this since I've only just recently migrated to Yarn [v1], but the .pnp.cjs file's packageLocation props have changed to my own dirs (e.g., from ./.yarn/cache/... to something like ../../../../../C:/Users/<USER>/AppData/Local/Yarn/Berry/cache/, is that what it's supposed to do?) since I did a yarn install. Won't start without running it so I had no choice.

I think I kinda get why this happened now. Did you happen to do a [Yarn Zero-Install](https://yarnpkg.com/features/zero-installs), @WarenGonzaga? In that case, I think you forgot to include the `./yarn/cache` in the repo so when I ran `yarn install`, `yarn add` it failed to find that and redownloaded/regenerated everything:

> The cache folder is by default stored within your project folder (in `.yarn/cache`). Just make sure you add it to your repository (see also, [Offline Cache](https://yarnpkg.com/features/offline-cache)).

**Comment by WarenGonzaga** [comment]
> > Lastly, I don't know much about this since I've only just recently migrated to Yarn [v1], but the .pnp.cjs file's packageLocation props have changed to my own dirs (e.g., from ./.yarn/cache/... to something like ../../../../../C:/Users//AppData/Local/Yarn/Berry/cache/, is that what it's supposed to do?) since I did a yarn install. Won't start without running it so I had no choice.
> 
> I think I kinda get why this happened now. Did you happen to do a [Yarn Zero-Install](https://yarnpkg.com/features/zero-installs), @WarenGonzaga? In that case, I think you forgot to include the `./yarn/cache` in the repo so when I ran `yarn install`, `yarn add` it failed to find that and redownloaded/regenerated everything:
> 
> > The cache folder is by default stored within your project folder (in `.yarn/cache`). Just make sure you add it to your repository (see also, [Offline Cache](https://yarnpkg.com/features/offline-cache)).

I'm trying to implement the Yarn's Zero Install but when I tried to apply gulp.js to the project it broke like hell. I guess the project is not ready for Zero Install (Plug n Play). I'm having an issue with that part.

**Comment by WarenGonzaga** [comment]
> **PR Includes**:
> 
> * Ignore .vscode folder
> * Add Prettier, config, ignore, scripts
> * Add ESLint, config, ignore, scripts
> * Add Yarn PnP VSCode integration
> * Update CONTRIBUTING.md
> * Add CI workflow
> * Some codestyle fixes
> 
> **Summary**:
> 
> Adds some checks and automations for anyone looking to contribute code (no testing yet.) The ESLint config extends the [eslint-config-google](https://github.com/google/eslint-config-google). Overwritten rules are of my own personal preference.
> 
> Also, since [Yarn PnP](https://yarnpkg.com/features/pnp) is used, some contributors that might be using VSCode with _Prettier_ and _ESLint_ extensions installed might encounter issues such as failing to find their respective modules so I [generated some editor SDKs for integration with VSCode](https://yarnpkg.com/getting-started/editor-sdks#vscode). These extensions should continue to work within this workspace. Basically, they fail to find a `node_modules` folder so they error out without this.
> 
> Lastly, I don't know much about this since I've only just recently migrated to Yarn [v1], but the `.pnp.cjs` file's `packageLocation` props have changed to my own dirs (e.g., from `./.yarn/cache/...` to something like `../../../../../C:/Users/<USER>/AppData/Local/Yarn/Berry/cache/`, is that what it's supposed to do?) since I did a `yarn install`. Won't start without running it so I had no choice.
> 
> **p.s.** Reminded me of those awesome ASCII rendering stuff, but this time with colors! So got a little excited and decided to help a little. Please do tell if there were some issues. Might add more PRs for new features/tests soon.

Huge thanks for the effort to improve the project, I really appreciate it. I'll review it now.

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,6 @@ yarn-error.log
 
 # parceljs
 .parcel-cache
+
+# VSCode user-specific settings
+/.vscode
```
Comment (WarenGonzaga) [inline .gitignore:20]: Add line break for this part. Also is this intended? ``/.vscode`` ? It is supposed to be ``.vscode`` only.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -12,7 +12,9 @@ There are many ways to contribute to this open source project. Any contributions
 
 ### 🧬 Development
 
-If you can write a code then create a pull request to this repo and I will review your code. Please consider to submit your pull request to the ```dev``` branch. I will auto reject if you submit your pull request to the ```main``` branch.
+If you can write a code then create a pull request to this repo and I will review your code. Please consider to submit your pull request to the `dev` branch. I will auto reject if you submit your pull request to the `main` branch.
+
+Before submitting a pull request, ensure that there is no code style issues by running `yarn run check`. If there are issues found, run `yarn run stylefix`.
```
Comment (WarenGonzaga) [inline CONTRIBUTING.md:17]: I like this, good catch!

File: src/index.js [hunk 3]
```diff
@@ -2,22 +2,22 @@
...
 
 document.head.appendChild(title);
 
 // apply correct repeat count to text
 for (let i = 0; i * textCount <= maxChar + textCount; i++) {
-    textPortrait.innerHTML = repeatText(data.text+' ', i);
-    console.log(data.text.length * i);
+  textPortrait.innerHTML = repeatText(data.text + " ", i);
+  console.log(data.text.length * i);
 }
```
Comment (WarenGonzaga) [inline src/index.js:23]: Let's use a single quote instead of a double quote.

File: src/scss/main.scss [hunk 4]
```diff
@@ -1,2 +1,2 @@
-@import 'vars'; // _vars (settings)
-@import 'base'; // _base (defaults)
\ No newline at end of file
+@import "vars"; // _vars (settings)
+@import "base"; // _base (defaults)
```
Comment (WarenGonzaga) [inline src/scss/main.scss:2]: Single quotes would work.

**Review by WarenGonzaga — CHANGES_REQUESTED** [review]
Thanks for the contribution, please review my requested changes. My workflow is mostly on single quotes and 4 spaces indents. I don't know what the standard is but from my recent popular projects we are using 2 space indent so I guess that's good for this also. My only concern is the single quotes implementation I guess you need to reconfigure the `prettier` I have config from the previous project I'll reference here so you can get an idea.

**Comment by WarenGonzaga** [comment]
Hey, @decanTyme this is what we use from my previous project.
https://github.com/warengonzaga/gathertown.js/blob/main/.prettierrc.json

Please use this as a reference.

File: .gitignore (hunk 1 above)
Comment (decanTyme) [reply .gitignore:20]: Ah. Sorry, a habit of mine. Somehow forgot those were dirs and not files. Looks like I got it everywhere, so I might have to change a few.

File: CONTRIBUTING.md (hunk 2 above)
Comment (decanTyme) [reply CONTRIBUTING.md:17]: Thanks! Thinking maybe the `check` script could run some tests too if It gets implemented.

**Comment by decanTyme** [comment]
> Thanks for the contribution, please review my requested changes. My workflow is mostly on single quotes and 4 spaces indents. I don't know what the standard is but from my recent popular projects we are using 2 space indent so I guess that's good for this also. 

Yeah, 2 indents are [probably the way to go for any Node.js projects](https://eslint.org/docs/rules/indent#enforce-consistent-indentation-indent). Looks cleaner to me at least.

> My only concern is the single quotes implementation I guess you need to reconfigure the `prettier` I have config from the previous project I'll reference here so you can get an idea.


> Hey, @decanTyme this is what we use from my previous project. https://github.com/warengonzaga/gathertown.js/blob/main/.prettierrc.json
> 
> Please use this as a reference.

Got that. I'll also revert back to a `.json` ESLint config since I only made it into a `.js` because of line ending issues. Some checks fail when run in Ubuntu, for example, when developing in win32 and config is set incorrectly. Got really confused for some time _why_ it fails. I didn't realize till now VSCode plays nicely with LF.

**Comment by WarenGonzaga** [comment]
> Got that. I'll also revert back to a `.json` ESLint config since I only made it into a `.js` because of line ending issues. Some checks fail when run in Ubuntu, for example, when developing in win32 and config is set incorrectly. Got really confused for some time _why_ it fails. I didn't realize till now VSCode plays nicely with LF.

Are you having issues right now?

**Comment by decanTyme** [comment]
> Are you having issues right now?

None so far with the code. Though sometimes Yarn SDKs get cranky and I have to regenerate or extensions won't work (again.) There's also Parcel sometimes not detecting changes in the `scss` files and builds on whatever the last change was. Currently working on something similar to #13, but for dynamically calculating `maxChar`. I'll head over there and give my thoughts.

> I'm trying to implement the Yarn's Zero Install but when I tried to apply gulp.js to the project it broke like hell. I guess the project is not ready for Zero Install (Plug n Play). I'm having an issue with that part.

I guess I can take a look at that too. Exporting builds looks interesting and I might learn a thing or two before I fully migrate to Yarn.

**Comment by decanTyme** [comment]
Please take a look at the new changes (916b413), should have covered all your requests. Let me know if there's still issues. If none, I will mark this as ready and I guess all that's left after that is to properly configure Yarn's Zero Install starting with including `./.yarn/cache` in the repo and reinstalling/regenerating deps as necessary.

**Comment by WarenGonzaga** [comment]
> None so far with the code. Though sometimes Yarn SDKs get cranky and I have to regenerate or extensions won't work (again.) There's also Parcel sometimes not detecting changes in the scss files and builds on whatever the last change was. Currently also working on something similar to #13, but for dynamically calculating maxChar since hardcoding it means it won't work for all screen sizes and orientations. Should this be a new issue? or should I give my thoughts in #13 directly?

One more note on this, the variable `$regular` in `_vars.scss` can directly affect the `maxChar` variable at `index.js`. Meaning whenever you change the font size the max character we set will have a direct impact on the output. I'd like to use a monotype font in the future so we can predictively compute how many characters can fit on the screen.

For the image resize, I don't think it can directly affect the `maxChar`? Since the image is currently being used as background.

> I guess I can take a look at that too. Exporting builds looks interesting and I might learn a thing or two before I fully migrate to Yarn.

The current build I guess is not Plug n Play I remember I reverted it back after I got errors. @decanTyme

**Comment by WarenGonzaga** [comment]
> @WarenGonzaga please take a look at the new changes ([916b413](https://github.com/WarenGonzaga/css-text-portrait-builder/commit/916b413d30493d9d6d21dff434fd4ef9d6f53f81)), should have covered all your requests. Let me know if there's still issues. If none, I will mark this as ready and I guess all that's left after that is to properly configure Yarn's Zero Install starting with including `./.yarn/cache` in the repo and reinstalling/regenerating deps as necessary.

Awesome, I'll review it again.

**Review by WarenGonzaga — APPROVED** [review]
For me, it is all good now.

**Comment by decanTyme** [comment]
> One more note on this, the variable `$regular` in `_vars.scss` can directly affect the `maxChar` variable at `index.js`. Meaning whenever you change the font size the max character we set will have a direct impact on the output. I'd like to use a monotype font in the future so we can predictively compute how many characters can fit on the screen.

Yep, I factored this one too. Exporting `scss` variables is possible and with Parcel it couldn't have been any easier. I also factored in getting font information at runtime so this shouldn't be a problem. Resize events shouldn't be an issue either. Animation would then be possible after I complete this work.

> The current build I guess is not Plug n Play I remember I reverted it back after I got errors. @decanTyme

Hmm. I guess it didn't roll back all the way, I still don't see any `node_modules` folder, or is Yarn v2 supposed to be like that?

Anyhow, we'll leave those for another issue/pr I guess.

**Comment by WarenGonzaga** [comment]
Thanks, @decanTyme I'll merge this now!

**Pull request merged**